docs/dev_guide/router_rest_spec.md

```diff
@@ -0,0 +1,270 @@
+---
+layout: default
+title: Router REST Spec
+parent: Developer Guides
+permalink: /docs/dev_guide/router_rest_spec
+---
+# Router REST Spec
+This page describes how the router's REST endpoints work for reading data out of Venice with the thin client. Currently,
+there exists a reference implementation of the thin client in Java, but if clients in additional languages are to be 
+built, they could reference this spec to do so.
+
+Note that although this part of the architecture is RESTful, Venice in general has a bias for performance, so the
+implementation may not be fully up to the standards of a dogmatic RESTafarian. For example, because of URL length
+limitations, batch gets are implemented with a POST rather than a GET, even though they carry no side effects.
+
+The following sections describe the endpoints in the order they are invoked by the client during initialization, in
+order to prime its internal state.
+
+## Cluster Discovery Endpoint
+The first endpoint the client invokes is to discover which cluster the store of interest belongs to:
+
+```
+GET https://<host>:<port>/discover_cluster/<store_name>
+```
+
+This will return a JSON response in the following format:
+
+```json
+{
+  "cluster": "cluster_name",
+  "name": "store_name",
+  "error": null,
+  "errorType": null,
+  "d2Service": "D2_SERVICE_NAME_62513b39957e_af278fc8",
+  "serverD2Service": "SERVER_D2_SERVICE_NAME_62513b3d13a5_4dc8aa80",
+  "zkAddress": "localhost:63359",
+  "kafkaBootstrapServers": "localhost:1234",
+  "exceptionType": null
+}
+```
+
+For all JSON responses returned by Venice, it is advisable to validate that the `error` field is `null`. If it is not
+`null`, then it should be a string containing some description of the error. In addition to that, the `errorType` field
+is a means to categorize errors programmatically according to the values in the 
+[ErrorType](https://github.com/linkedin/venice/blob/main/internal/venice-client-common/src/main/java/com/linkedin/venice/exceptions/ErrorType.java) 
+enum.
+
+The client then uses the information in the `d2Service` field of this response to connect to the correct cluster.
+
+A few important details:
+
+1. The cluster discovery mechanism currently makes certain assumptions that 
+   [D2](https://linkedin.github.io/rest.li/Dynamic_Discovery) is used. This assumption could be loosened in the future 
+   in order to integrate with other service discovery mechanisms.
+2. This step can be skipped if not leveraging multi-cluster deployments, or if the operator wishes to let clients 
```

Comment:
Shall we mention that the customer needs to design some basic load-rebalancing mechanism when hard-coding cluster address, such as round-robin otherwise, some Router will be heavily used? D2 offers this natively.

Reply:
If the non-Java clients want to be more performant, some connection pool management will be required.

Reply:
Another important thing is that the non-Java clients need to provide correct certificate to pass ACL check.

Reply:
These are all valid points, though they do not seem Venice-specific, right? A a client to any system will need to have load balancing, connection pooling, auth/auth...

Reply:
I added a paragraph about it in my latest commit.